# Exploring eBay Car Sales Data

In this guided project, we worked with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The data comprises of 50,000 sampled data points from the full dataset.

## Part1:  Import packages and understand the data

In [1]:
# Importing packages and reading in the csv data
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
# checking the preview of the autos table
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

The dataset contains 5000 rows of information with 20 columns

In [3]:
# Checking the informatin about the autos table
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Part 2: Convert column names to snakecase and rename some column names based on data dictionary

### Array of column names to be used to make edits

In [4]:
# Making a copy of the column names
old_columns = autos.columns.copy()

In [5]:
# Prapare an array of column names in snakecase format
new_columns = [
                'date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
                'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
                'kilometers', 'registration_month', 'fuel_type', 'brand',
                'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
                'last_seen'
               ]

In [6]:
# Renaming the column names
autos.columns = new_columns

In [7]:
# Checking that the new columns names are properly changed
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model kilometers  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

**The following are the new column names that will be used from now on in the dataframe. The column names were changed to that they conform to the snakecase format from the original camelcase in which they were.**

## Part 3: Basic Data Exploration and Cleaning Tasks

In [8]:
# Looking at statistics for the complete autos table (both numerical and categorical data)
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       kilometers  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

**The columns _seller_, _offer_type_ and _abtest_ have each mostly one value. These columns need to be investigated further.**

In [9]:
# Checking the unique values for 'seller' column in autos table
autos["seller"].unique()

array(['privat', 'gewerblich'], dtype=object)

In [10]:
# Determine number of unique values from 'seller' column in autos table that is equal to 'privat'
# There are 49999 out of 50000 items that are 'private'
autos.loc[autos['seller'] == 'privat', 'seller'].value_counts()

privat    49999
Name: seller, dtype: int64

In [11]:
# Checking the unique values for 'offer_type' column in autos table
autos["offer_type"].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [12]:
# Determine number of unique values from 'offer_type' column in autos table that is equal to 'Angebot'
# There are 49999 out of 50000 items that are 'Angebot'
autos.loc[autos['offer_type'] == 'Angebot', 'offer_type'].value_counts()

Angebot    49999
Name: offer_type, dtype: int64

**The two columns _seller_ and _offer type_ can be dropped as they only have mostly one type of information.**

In [13]:
# Checking the datatypes for the columns in autos table
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_PS               int64
model                 object
kilometers            object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

**The infomation in columns _price_, _kilometers_ should be integers while _ad_created_ and _last seen_ are dates.**

In [14]:
# Previewing some values in the 'price' column of the autos table
autos["price"].tail()

49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, dtype: object

**For _price_, the dollar symbol and the coma is to be removed, type changed to integer and <br>
column name _price_ changed to _price_usd_**

In [16]:
# This code removes the '$' and ',' and changes datatype to integer in the 'price' column
autos['price'] = autos['price'].str.replace('$', "")\
                               .str.replace(",", "")\
                               .astype(int)

In [17]:
# This renames the 'price' column to 'price_usd' and commits the change to autos table
autos.rename(columns={'price': 'price_usd'}, inplace=True)

In [18]:
# Checking to see that the changes are properly effected
autos["price_usd"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_usd, dtype: int64

In [19]:
# Previewing some values in the 'kilometers' column of the autos table
autos["kilometers"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: kilometers, dtype: object

**For _kilometers_, the km symbol and the coma is to be removed and type changed to integer**

In [20]:
# This removes the 'km' and ',' and changes datatype to integer in 'kilometers' column
autos['kilometers'] = autos['kilometers'].str.replace('km', "")\
                               .str.replace(",", "")\
                               .astype(int)

In [21]:
# Checking to see that the changes are properly effected
autos["kilometers"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: kilometers, dtype: int64

## Part 4: Further Data Exploration - Removing Outliers

In [25]:
# Checking the number of unique values in 'kilometers' column
# Checking the 'kilometers' statistics
print(autos["kilometers"].unique().shape)
autos["kilometers"].describe()

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometers, dtype: float64

In [26]:
# Checking the distribution of the unique values in the 'kilometers' column
# The distribution look to be fine, no need to change anyting here
autos["kilometers"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: kilometers, dtype: int64

In [27]:
# Checking the number of unique values in 'price_usd' column
# Checking the 'price_usd' statistics
print(autos["price_usd"].unique().shape)
autos['price_usd'].describe().apply(lambda x: format(x, 'f'))

(2357,)


count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price_usd, dtype: object

In [36]:
# Checking to see the maximum threshold value i.e. 999990
autos['price_usd'].sort_values(ascending=False).head(20)

39705    99999999
42221    27322222
27371    12345678
39377    12345678
47598    12345678
2897     11111111
24384    11111111
11137    10000000
47634     3890000
7814      1300000
22947     1234566
43049      999999
514        999999
37585      999990
36818      350000
14715      345000
34723      299000
35923      295000
12682      265000
47337      259000
Name: price_usd, dtype: int64

In [66]:
# Checking to see the minimum threshold value i.e. 100
autos['price_usd'].sort_values(ascending=False)[-1940:-1890]

37393    120
28929    120
12848    120
13145    120
21902    120
39136    120
42051    120
29936    120
25652    120
15144    120
6057     120
7439     120
14150    120
47523    120
46675    120
29813    120
18072    120
35561    120
13134    120
23702    120
9318     120
42526    120
42869    120
30688    120
31602    120
40847    120
17506    120
15309    120
3468     120
43161    120
49698    120
36534    120
28057    120
4176     120
37577    120
17768    120
12790    117
46769    115
45992    115
29150    111
39221    111
48686    110
41730    110
12452    110
41898    100
34015    100
34951    100
30286    100
2557     100
35241    100
Name: price_usd, dtype: int64

**All prices above 900000 and below 100 does not seem reasonable for the price of the cars. So these can be removed**

In [67]:
# Removing all the prices outside the range 100 to 900000
autos = autos[autos["price_usd"].between(100,900000)]
autos.shape

(48224, 20)

In [68]:
# Checking the statistics after removing the outliers in the 'price_usd' column
autos["price_usd"].describe().apply(lambda x: format(x, "f"))

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price_usd, dtype: object

In [69]:
autos["price_usd"].value_counts().sort_index(ascending=False).head()

350000    1
345000    1
299000    1
295000    1
265000    1
Name: price_usd, dtype: int64

**The remaining values are better now, after having removed the outliers.**

## Part 5: Data Exploration of Date Columns

In [70]:
# Previewing the frist 5 values of the 3 named columns of the autos table
# The are all date columns
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [71]:
# This code picks only the first 10 string characters of the date for 'date_crawled' column
# The second part calculates the percentage of each date out of the total number
# and sorts the dates in ascending order 
autos.loc[:, "date_crawled"] = autos.loc[:, "date_crawled"].str[:10]
autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

**The data was crawled between 3rd March to 7 April 2016 for each day. The distribution of most of the values are ranging between 1 to 3 percent of the total, except for data that was collected on 6 and 7 April, which was less than 1 percent.**

In [72]:
# This code picks only the first 10 string characters of the date for'ad_created' column
# The second part calculates the percentage of each date out of the total number
# and sorts the dates in ascending order 
pd.options.display.max_rows = 100
autos.loc[:, "ad_created"] = autos.loc[:, "ad_created"].str[:10]
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
2016-02-22    0.000021
2016-02-23    0.000083
2016-02-24    0.000041
2016-02-25    0.000062
2016-02-26    0.000041
2016-02-27    0.000124
2016-02-28    0.000207
2016-02-29    0.000166
2016-03-01    0.000104
2016-03-02    0.000104
2016-03-03    0.000871
2016-03-04    0.001493
2016-03-05    0.022914
2016-03-06 

**The ads were created between 11 June 2015 to 7 April 2016. Adds created appear to start off slow from June 2015 then increase up in March 2016, remain stable for much of the time until begining of April 2016 then slow down again from there.**

In [73]:
# This code picks only the first 10 string characters of the date for'last_seen' column
# The second part calculates the percentage of each date out of the total number
# and sorts the dates in ascending order
autos.loc[:, "last_seen"] = autos.loc[:, "last_seen"].str[:10]
autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005433
2016-03-08    0.007320
2016-03-09    0.009580
2016-03-10    0.010638
2016-03-11    0.012400
2016-03-12    0.023785
2016-03-13    0.008875
2016-03-14    0.012629
2016-03-15    0.015863
2016-03-16    0.016444
2016-03-17    0.028098
2016-03-18    0.007320
2016-03-19    0.015760
2016-03-20    0.020654
2016-03-21    0.020550
2016-03-22    0.021359
2016-03-23    0.018580
2016-03-24    0.019762
2016-03-25    0.019098
2016-03-26    0.016672
2016-03-27    0.015552
2016-03-28    0.020840
2016-03-29    0.022292
2016-03-30    0.024697
2016-03-31    0.023826
2016-04-01    0.022852
2016-04-02    0.024884
2016-04-03    0.025133
2016-04-04    0.024531
2016-04-05    0.125062
2016-04-06    0.221964
2016-04-07    0.132154
Name: last_seen, dtype: float64

**The _last_seen_ date information was collected on a daily basis between 3 March to 7 April 2016 and looks uniformly distributed.**

## Part 6: Counting Number of Listings Outside the 1900 to 2016 Interval

In [76]:
# Previewing the first 5 items in the 'registration_year' column
autos["registration_year"].head()

0    2004
1    1997
2    2009
3    2007
4    2003
Name: registration_year, dtype: int64

In [78]:
# Checking statistics for the 'registration_year' column
autos["registration_year"].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**The registration year of 1900 (too old) and that of 9999 (outlier) seems to be off and possibly a mistake. I determine that the values outside the range of 1900 and 2016 should be removed.**

In [79]:
# Determining number of listings with cars that fall outside 
# the 1900 - 2016 interval
print(autos[(autos["registration_year"] < 1900) | 
            (autos["registration_year"] > 2016)].shape)
autos.shape

(1872, 20)


(48224, 20)

**1872 out of 48224 listings fall outside the range 1900 and 2016. This accounts for approximately 4% and can safely be removed.**

In [80]:
# Removing the items that fall outside the range 1900 to 2016 of registration years
# Calculating percentage of each 'registration_year' out of the total number
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=True)

1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
1937    0.000086
1938    0.000022
1939    0.000022
1941    0.000043
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000043
1952    0.000022
1953    0.000022
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000475
1961    0.000129
1962    0.000086
1963    0.000173
1964    0.000259
1965    0.000367
1966    0.000475
1967    0.000561
1968    0.000561
1969    0.000410
1970    0.000798
1971    0.000561
1972    0.000712
1973    0.000496
1974    0.000518
1975    0.000388
1976    0.000453
1977    0.000475
1978    0.000906
1979    0.000734
1980    0.001747
1981    0.000582
1982    0.000885
1983    0.001100
1984    0.001100
1985    0.002006
1986    0.001532
1987    0.001553
1988    0.002869
1989    0.003689
1990    0.007163
1991    0.007292
1992    0.007918
1993    0.009061
1994    0.013505
1995    0.025738
1996    0.029233
1997    0.0415

**There is a steady increase in registration from 1910 and it peaks around 1996 to 2009, and then drops from then on.**

## Part 7: Aggregation Analysis

In [81]:
# Checking the unique items in the 'brand' column of the autos table
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [83]:
autos['brand'].value_counts()

volkswagen        9799
bmw               5107
opel              4971
mercedes_benz     4480
audi              4022
ford              3237
renault           2182
peugeot           1384
fiat              1187
seat               846
skoda              761
nissan             711
mazda              706
smart              658
citroen            651
toyota             593
hyundai            464
sonstige_autos     442
volvo              423
mini               408
mitsubishi         379
honda              365
kia                328
alfa_romeo         309
porsche            279
suzuki             275
chevrolet          263
chrysler           163
dacia              123
daihatsu           116
jeep               106
land_rover          98
subaru              98
saab                77
jaguar              71
daewoo              69
trabant             63
rover               62
lancia              49
lada                27
Name: brand, dtype: int64

**I decide to only agregate by brands that contribute over 5% of total number. These are Volkswagen, BMW, Mercedes Benz, Opel, Audi and Ford**

In [84]:
# Determine percentage contribution of each brand out of the total number
brands = autos["brand"].value_counts(normalize=True, sort=True)
# Select the brands whose contribution is greater or equal to 5% of total number
selected_brands = brands[brands>=0.05]
selected_brands

volkswagen       0.211404
bmw              0.110179
opel             0.107245
mercedes_benz    0.096652
audi             0.086771
ford             0.069835
Name: brand, dtype: float64

In [91]:
# Create an empty dictionary to store the aggregated data
brand_data = {}

# Use a for loop to iterate over the selected brands
for brand in selected_brands.index:
    # Assign the mean price to the dictionary with the brand name as key
    selected_rows = autos[autos["brand"]==brand]
    meanprice = selected_rows['price_usd'].mean()
    # Print out the dictionary of aggregated data
    brand_data[brand] = meanprice

In [93]:
# initializing K for rounding off
K = 3
# loop to iterate for values 
res = dict()
for key in brand_data:
    # rounding to K using round()
    res[key] = round(brand_data[key], K)
#Rounded result 
res

{'volkswagen': 5436.95,
 'bmw': 8381.677,
 'opel': 3005.496,
 'mercedes_benz': 8672.654,
 'audi': 9380.719,
 'ford': 3779.269}

**From the top 5% brands, Audi is the most expensive, followed by Mercedes Benz and BMW. Ford and Opel are the cheapest among the top 5% and Volkwagen is in the middle**

## Part 8: Combine Data from Two Series into Single Dataframe

In [89]:
# Checking first 5 items in 'kilometers' column
autos["kilometers"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: kilometers, dtype: int64

In [94]:
# Retrieving the top 5% brand labels
brands = autos["brand"].value_counts(normalize=True, sort=True)
brand_labels = brands[brands >= 0.05].index
brand_labels

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [98]:
# Calculating the brand mean prices for the top 5% brands
brand_mean_prices = {}
for brand in brand_labels:
    price = autos[autos["brand"] == brand]['price_usd']
    brand_mean_prices[brand] = int(price.mean())
brand_mean_prices

{'volkswagen': 5436,
 'bmw': 8381,
 'opel': 3005,
 'mercedes_benz': 8672,
 'audi': 9380,
 'ford': 3779}

In [99]:
# Calculating the brand mean kilometers for the top 5% brands
brand_mean_kilos = {}
for brand in brand_labels:
    kilos = autos[autos["brand"] == brand]['kilometers']
    brand_mean_kilos[brand] = int(kilos.mean())
brand_mean_kilos

{'volkswagen': 128799,
 'bmw': 132695,
 'opel': 129384,
 'mercedes_benz': 131025,
 'audi': 129245,
 'ford': 124277}

In [107]:
# Create brand mean price (bmp) series table
bmp_series = pd.Series(brand_mean_prices, name='mean_price')

# Create brand mean kilos (bmk) series table
bmk_series = pd.Series(brand_mean_kilos, name='mean_km')

# Concatenate bmp_series and bmk_series into a dataframe table df
# Sort the table by mean_price and mean_km and arrange in descending order
df = pd.concat([bmp_series, bmk_series], axis=1)
df.sort_values(by=['mean_price', 'mean_km'], ascending=False)

mean_price  mean_km
audi                 9380   129245
mercedes_benz        8672   131025
bmw                  8381   132695
volkswagen           5436   128799
ford                 3779   124277
opel                 3005   129384

**The three more expensive brands, Audi, Mecedes benz and BMW also have the highest mileage according to this analysis.**

## Part 9: Translate German words into English

In [108]:
# Listing the autos column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_usd', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'kilometers', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [257]:
# Checking the unique values for 'seller' column
autos["seller"].unique()

array(['privat', 'gewerblich'], dtype=object)

In [109]:
# Create a dictionary used to translate 'seller' column values from Germany to English
seller = {"privat":"private", "gewerblich":"commercial"}
autos["seller"] = autos["seller"].map(seller)
autos["seller"].unique()

array(['private', 'commercial'], dtype=object)

In [111]:
# Checking unique values for the 'vehicle_type' column
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'compact', 'station wagon', nan, 'coupe',
       'suv', 'convertible', 'other'], dtype=object)

In [110]:
# Create a dictionary used to translate 'vehicle_type' column values from Germany to English
vehicle = {"bus":"bus", "limousine":"limousine", "kleinwagen": "compact",
           "kombi": "station wagon", "coupe": "coupe", "suv": "suv",
           "cabrio": "convertible", "andere": "other"}
autos["vehicle_type"] = autos["vehicle_type"].map(vehicle)
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'compact', 'station wagon', nan, 'coupe',
       'suv', 'convertible', 'other'], dtype=object)

In [263]:
# Checking unique values for 'unrepaired_damage' column
autos["unrepaired_damage"].unique()

array(['nein', nan, 'ja'], dtype=object)

In [112]:
# Create a dictionary used to translate 'unrepaired_damage' column values from Germany to English
damage = {"nein":"no", "ja": "yes"}
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(damage)
autos["unrepaired_damage"].unique()

array(['no', nan, 'yes'], dtype=object)

## Most Common Brand/Model Combinations

In [113]:
autos["kilometers"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: kilometers, dtype: int64

In [115]:
# Calculate the model counts grouped by vehicle brand type
# Sort the resulting table indescing order and show only the first 10 items
gp = autos.groupby('brand')['model'].value_counts().sort_values(ascending=False)[0:10]
gp

brand          model   
volkswagen     golf        3684
bmw            3er         2602
volkswagen     polo        1592
opel           corsa       1568
volkswagen     passat      1345
opel           astra       1337
audi           a4          1226
mercedes_benz  c_klasse    1135
bmw            5er         1123
mercedes_benz  e_klasse     953
Name: model, dtype: int64

**Volkswagen (Golf, Polo, Passat) BMW (3ER, 5ER), Opel (Corsa, Astra) and Mercedes Benz (E Class) are among the most common brands**

## Average Price compared to Mileage

In [129]:
# Calculate the mean prices grouped by kilometers
autos.groupby('kilometers')['price_usd'].mean().sort_values(ascending=False).round(1)

kilometers
10000     21255.3
20000     18523.2
30000     16674.6
40000     15519.0
50000     13853.9
60000     12407.0
70000     10954.8
5000       9794.1
80000      9743.1
90000      8470.1
100000     8188.2
125000     6238.3
150000     3792.3
Name: price_usd, dtype: float64

**It can be seen that the mean price generally reduces as the mileage increases**

## Brand Compared to Unrepaired Damage

In [117]:
df

mean_price  mean_km
volkswagen           5436   128799
bmw                  8381   132695
opel                 3005   129384
mercedes_benz        8672   131025
audi                 9380   129245
ford                 3779   124277

In [121]:
# Convert 'df' dataframe to a list (top_brands)
top_brands = df.index.to_list()

# Select only the brands from autos dataframe that are in the top_brands list
top_autos = autos[autos['brand'].isin(top_brands)]
top_brands

['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford']

In [130]:
pd.options.display.max_rows = 200

# Determine the mean price for each of the top brands grouped based on
# whether they have unrepaired damage or not
damage_cost = top_autos.groupby(['brand', 'unrepaired_damage'])['price_usd'].mean().round(1)
damage_cost

brand          unrepaired_damage
audi           no                   10929.2
               yes                   3324.7
bmw            no                    9454.4
               yes                   3538.9
ford           no                    4666.7
               yes                   1408.3
mercedes_benz  no                    9815.4
               yes                   3933.3
opel           no                    3672.3
               yes                   1397.0
volkswagen     no                    6483.2
               yes                   2194.6
Name: price_usd, dtype: float64

In [132]:
damage_df = pd.DataFrame(damage_cost)
damage_df

price_usd
brand         unrepaired_damage           
audi          no                   10929.2
              yes                   3324.7
bmw           no                    9454.4
              yes                   3538.9
ford          no                    4666.7
              yes                   1408.3
mercedes_benz no                    9815.4
              yes                   3933.3
opel          no                    3672.3
              yes                   1397.0
volkswagen    no                    6483.2
              yes                   2194.6

**It is evident from this information about the top brands that brands without unrepaired damage cost about 1/2 to 1/3 the price of brands with repaired  damage**